File: backend/src/routes/team.js
import express from "express";
import { requireAuth, requireRole } from "../middleware/auth.js";
import { ROLES } from "../constants/roles.js";
import { getTeams, getTeamMembers, getAllUsers } from "../services/userService.js";
import { getUserMealStatus } from "../services/mealService.js";
import { getAvailableMeals } from "../services/availabilityService.js";
import { getEffectiveLocation } from "../services/workLocationService.js";

const router = express.Router();

router.get("/", requireAuth, (req, res) => {
    res.json(getTeams());
});

/**
 * GET /api/team/participation?date=YYYY-MM-DD
 * TL: own team members with meal/location status.
 * Admin: all users with meal/location status.
 * Logistics: team-level aggregates only, no individual names.
 */
router.get(
    "/participation",
    requireAuth,
    requireRole([ROLES.TEAM_LEAD, ROLES.ADMIN, ROLES.LOGISTICS]),
    (req, res) => {
        const user = req.session.user;
        const date = req.query.date;

        if (user.role === ROLES.LOGISTICS) {
            const teams = getTeams();
            const allUsers = getAllUsers();
            const result = teams.map((t) => ({
                teamId: t.id,
                teamName: t.name,
                memberCount: allUsers.filter((u) => u.teamId === t.id).length,
            }));
            return res.json(result);
        }

        // TL sees own team; Admin sees all
        let members;
        if (user.role === ROLES.ADMIN) {
            members = getAllUsers();
        } else {
            members = getTeamMembers(user.teamId);
        }

        // If date provided, include meal status and location per member
        if (date) {
            const availableMeals = getAvailableMeals(date);

            const result = members.map((m) => ({
                id: m.id,
                name: m.name,
                role: m.role,
                teamId: m.teamId,
                location: getEffectiveLocation(m.id, date),
                meals: getUserMealStatus(m.id, availableMeals, date),
            }));
            return res.json(result);
        }

        // No date: return member list only
        const safe = members.map((m) => ({
            id: m.id,
            name: m.name,
            role: m.role,
            teamId: m.teamId,
        }));

        res.json(safe);
    }
);

export default router;

File: mhp-python/app/routes/team.py
from fastapi import APIRouter, Depends

from app.auth import get_current_user
from app.constants import Roles
from app.services.user_service import get_all_teams, get_all_users, get_team_members
from app.services.availability_service import get_available_meals
from app.services.meal_service import get_user_meal_status
from app.services.work_location_service import get_effective_location

router = APIRouter()


@router.get("/")
def list_teams(user: dict = Depends(get_current_user)):
    return get_all_teams()


@router.get("/participation")
def team_participation(date: str, user: dict = Depends(get_current_user)):
    """
    Returns per-user participation for a date, scoped by role:
    - ADMIN: all users
    - TEAM_LEAD: own team only
    - LOGISTICS: all users but names are hidden (aggregated view)
    - EMPLOYEE: 403 (no access)
    """
    from fastapi import HTTPException
    if user["role"] == Roles.EMPLOYEE:
        raise HTTPException(status_code=403, detail="Employees cannot view team participation")

    available = get_available_meals(date)

    if user["role"] == Roles.TEAM_LEAD:
        users = get_team_members(user["teamId"])
    else:
        users = get_all_users()

    result = []
    for u in users:
        meal_status = get_user_meal_status(u["id"], available, date)
        location = get_effective_location(u["id"], date)
        entry = {
            "userId": u["id"],
            "name": u["name"] if user["role"] != Roles.LOGISTICS else "—",
            "teamId": u.get("teamId"),
            "location": location,
            "meals": meal_status,
        }
        result.append(entry)

    return {"date": date, "users": result}
